src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1343,25 +1365,46 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep, uct_pending_purge_callback_t c
     }
 }
 
-ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
+ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface,
+                                     uct_dc_mlx5_ep_t *ep)
 {
+    ucs_status_t ret_status = UCS_OK;
     ucs_status_t status;
 
     if (iface->super.super.config.fc_enabled) {
-        UCT_RC_CHECK_FC_WND(&ep->fc, ep->super.stats);
-        if (ep->fc.fc_wnd == iface->super.super.config.fc_hard_thresh) {
+        if (ucs_unlikely((ep->fc.fc_wnd ==
+                          iface->super.super.config.fc_hard_thresh) ||
+                         (ep->fc.fc_wnd <= 0))) {
+            if (ep->fc.fc_wnd <= 0) {
+                UCS_STATS_UPDATE_COUNTER(ep->fc.stats, UCT_RC_FC_STAT_NO_CRED, 1);
+                UCS_STATS_UPDATE_COUNTER(ep->super.stats, UCT_EP_STAT_NO_RES, 1);
+                ret_status = UCS_ERR_NO_RESOURCE;
+            }
+
             status = uct_rc_fc_ctrl(&ep->super.super,
                                     UCT_RC_EP_FLAG_FC_HARD_REQ,
                                     NULL);
             if (status != UCS_OK) {
+                if ((ep->dci != UCT_DC_MLX5_EP_NO_DCI) && 
+                    !uct_dc_mlx5_iface_is_dci_rand(iface)) {
+                    /* Allow owning DCI if FC window is empty and EP sends
+                     * FC_HARD_REQ or EP has outstanding operations */
+                    ucs_assertv_always((ep->fc.fc_wnd <= 0) ||
```

Comment:
otherwise - the assertion fails, since DCI is owned by EP, but fc_ctrl function returns `UCS_ERR_NO_RESOURCE`

Reply:
why is DCI owned by ep, if there were not available resources to send grant?

Reply:
there are resources and EP sent `FC_HARD_REQ`, but it returns `UCS_ERR_NO_RESOURCE` in case of res-sending `FC_HARD_REQ` to make sure user won't send anything due to lack of resources

Reply:
If the transfer of previous hard request was completed, dc should be release. If it is still in progress, we have to see outstandings

Reply:
> If the transfer of previous hard request was completed, dci should be released

yes, it happens
but here we are sending new FC_HARD_REQ, since we don't receive FC_PURE_GRANT in a given time, but FC window is 0

Reply:
why would we hold dci in this case?